styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/TimeoutHandler.java

```diff
@@ -0,0 +1,85 @@
+/*-
+ * -\-\-
+ * Spotify Styx Scheduler Service
+ * --
+ * Copyright (C) 2016 - 2019 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.state.handlers;
+
+import static com.spotify.styx.state.StateUtil.hasTimedOut;
+
+import com.spotify.styx.model.Event;
+import com.spotify.styx.model.Workflow;
+import com.spotify.styx.model.WorkflowConfiguration;
+import com.spotify.styx.model.WorkflowInstance;
+import com.spotify.styx.state.OutputHandler;
+import com.spotify.styx.state.RunState;
+import com.spotify.styx.state.StateManager;
+import com.spotify.styx.state.TimeoutConfig;
+import com.spotify.styx.storage.Storage;
+import com.spotify.styx.util.Time;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link OutputHandler} that issues {@code timeout} events for instances that have timed out according to the
+ * {@link TimeoutConfig} and the {@link WorkflowConfiguration#runningTimeout()}.
+ */
+public class TimeoutHandler implements OutputHandler {
+
+  private static final Logger log = LoggerFactory.getLogger(TimeoutHandler.class);
+
+  private final TimeoutConfig ttls;
+  private final Time time;
+  private final StateManager stateManager;
+  private final Storage storage;
+
+  public TimeoutHandler(TimeoutConfig ttls, Time time, StateManager stateManager,
+                        Storage storage) {
+    this.ttls = Objects.requireNonNull(ttls, "ttls");
+    this.time = Objects.requireNonNull(time, "time");
+    this.stateManager = Objects.requireNonNull(stateManager, "stateManager");
+    this.storage = Objects.requireNonNull(storage, "storage");
+  }
+
+  @Override
+  public void transitionInto(RunState runState) {
+    // TODO: cache the workflow ttl configuration
```

Comment:
This seems pretty heavy. We should probably already do it in this PR.

We can actually pass down `running_timeout` to `RunState` to avoid reading the complete workflow. But that would mean users will not be able to change the value for an active state. :/

Or we can perhaps use the workflow cache in `StyxScheduler` for non critical path like this.

Reply:
Making the timeout handler use the workflow cache from the `StyxScheduler`. Also changing the `CachedSupplier` to avoid redundant fetch/compute on concurrent calls (which we'll have a lot of).